### PR TITLE
Deprecate vectorized methods for univariate distributions

### PR DESCRIPTION
**Pull request by simonbyrne**
This keeps the old behaviour for multivariate distributions, as I don't have any great ideas for those.

Still need to overload `Base.broadcast` for fast `UnitRange` methods (e.g. `Binomial`).

### CODE REVIEW

File: src/univariate/discrete/binomial.jl [hunk 1]
```diff
@@ -118,7 +118,9 @@ end
 RecursiveBinomProbEvaluator(d::Binomial) = RecursiveBinomProbEvaluator(d.n, d.p / (1 - d.p))
 nextpdf(s::RecursiveBinomProbEvaluator, pv::Float64, x::Integer) = ((s.n - x + 1) / x) * s.coef * pv
 
-function _pdf!(r::AbstractArray, d::Binomial, X::UnitRange)
+function Base.broadcast!(::typeof(pdf), r::AbstractArray, d::Binomial, X::UnitRange)
+    indices(r) == indices(X) || throw(ArgumentError("Inconsistent array dimensions."))
```
Comment (ararslan) [inline src/univariate/discrete/binomial.jl]: Perhaps `DimensionMismatch` would be more appropriate than `ArgumentError`?

File: src/univariate/discrete/betabinomial.jl [hunk 2]
```diff
@@ -81,8 +81,9 @@ function kurtosis(d::BetaBinomial)
     return (left * right) - 3
 end
 
-function pdf(d::BetaBinomial, k::Int)
+function pdf(d::BetaBinomial{T}, k::Int) where T
     n, α, β = d.n, d.α, d.β
+    0 <= k <= n || return zero(T)
```
Comment (ararslan) [inline src/univariate/discrete/betabinomial.jl]: Perhaps `insupport`?

Comment (simonbyrne) [inline src/univariate/discrete/betabinomial.jl]: I always feel that the arguments to `insupport` are the wrong way around.

File: src/univariates.jl [hunk 3]
```diff
@@ -551,8 +522,8 @@ function _pdf!(r::AbstractArray, d::DiscreteUnivariateDistribution, X::UnitRange
     return r
 end
 
-
-pdf(d::DiscreteUnivariateDistribution) = isbounded(d) ? pdf(d, minimum(d):maximum(d)) :
+# TODO: should we get rid of this?
```
Comment (ararslan) [inline src/univariates.jl]: Makes sense to me to get rid of it, maybe deprecate it for `pdf.(d, support(d))`?

File: src/univariate/discrete/betabinomial.jl (hunk 2 above)
Comment (ararslan) [inline src/univariate/discrete/betabinomial.jl]: Oh really? The current order makes sense to me. I think of it like `contains(haystack, needle)` from Base.

File: src/univariates.jl (hunk 3 above)
Comment (simonbyrne) [inline src/univariates.jl]: ok

Comment (ararslan) [inline src/univariates.jl]: If the user calls it on a distribution that doesn't have finite support, that's on them. 😛

File: src/univariate/discrete/betabinomial.jl (hunk 2 above)
Comment (simonbyrne) [inline src/univariate/discrete/betabinomial.jl]: well, it's the opposite of `in` for example.

File: src/univariate/discrete/betabinomial.jl [hunk 4]
```diff
@@ -81,29 +81,18 @@ function kurtosis(d::BetaBinomial)
...
-    return values
-end
-
-entropy(d::BetaBinomial) = entropy(Categorical(pdf(d)))
-median(d::BetaBinomial) = median(Categorical(pdf(d))) - 1
-mode(d::BetaBinomial{T}) where {T<:Real} = indmax(pdf(d)) - one(T)
-modes(d::BetaBinomial) = [x - 1 for x in modes(Categorical(pdf(d)))]
+entropy(d::BetaBinomial) = entropy(Categorical(pdf.(d,support(d))))
+median(d::BetaBinomial) = median(Categorical(pdf.(d,support(d)))) - 1
+mode(d::BetaBinomial{T}) = indmax(pdf.(d,support(d))) - 1
```
Comment (rofinn) [inline src/univariate/discrete/betabinomial.jl]: Forgot the `where {T}`

Comment (simonbyrne) [inline src/univariate/discrete/betabinomial.jl]: Yes, not really necessary.

Comment (ararslan) [inline src/univariate/discrete/betabinomial.jl]: I'm confused by that response. This should either have `where {T<:Real}` or (even better) have `BetaBinomial{T}` changed to `BetaBinomial`.

**Comment by simonbyrne** [comment]
Okay, this should be good to go.

File: src/univariate/discrete/betabinomial.jl (hunk 4 above)
Comment (simonbyrne) [inline src/univariate/discrete/betabinomial.jl]: ah, I missed that. Good point.

**Pull request merged**